adminpage/views.py

```diff
@@ -1,3 +1,272 @@
+            if activity.name != self.input['name']:
+                raise LogicError('Cannot modify the name of a published activity ')
+            if activity.place != self.input['place']:
+                raise LogicError('Cannot modify the place of a published activity')
+            if activity.book_start != (datetime.strptime(self.input['bookStart'], "%Y-%m-%dT%H:%M:%S.%fZ") + timedelta(hours=8)):
+                raise LogicError('Cannot modify the bookStart time of a published activity')
+            if self.input['status'] != 1:
+                raise LogicError('you cannot stage published activity')
+
+        if int(time.mktime(activity.end_time.timetuple())) < int(time.time()):
+            if activity.start_time != (datetime.strptime(self.input['startTime'], "%Y-%m-%dT%H:%M:%S.%fZ") + timedelta(hours=8)):
+                raise LogicError('Cannot modify startTime of a activity ended')
+            if activity.end_time != (datetime.strptime(self.input['endTime'], "%Y-%m-%dT%H:%M:%S.%fZ") + timedelta(hours=8)):
+                raise LogicError('Cannot modify endTime of a activity ended')
+
+        if int(time.mktime(activity.start_time.timetuple())) < int(time.time()):
+            if activity.book_end != (datetime.strptime(self.input['bookEnd'], "%Y-%m-%dT%H:%M:%S.%fZ") + timedelta(hours=8)):
+                raise LogicError('Cannot modify bookEnd time of a activity started')
+
+        if int(time.mktime(activity.book_start.timetuple())) < int(time.time()):
+            if activity.total_tickets != self.input['totalTickets']:
+                raise LogicError('Cannot modify totalTickets after book_start')
+
+        activity.name = self.input['name']
+        activity.place = self.input['place']
+        activity.description = self.input['description']
+        activity.pic_url = self.input['picUrl']
+        activity.start_time = (datetime.strptime(self.input['startTime'], "%Y-%m-%dT%H:%M:%S.%fZ") + timedelta(hours=8))
+        activity.end_time = (datetime.strptime(self.input['endTime'], "%Y-%m-%dT%H:%M:%S.%fZ") + timedelta(hours=8))
+        activity.book_start = (datetime.strptime(self.input['bookStart'], "%Y-%m-%dT%H:%M:%S.%fZ") + timedelta(hours=8))
+        activity.book_end = (datetime.strptime(self.input['bookEnd'], "%Y-%m-%dT%H:%M:%S.%fZ") + timedelta(hours=8))
+        activity.total_tickets = self.input['totalTickets']
+        activity.status = self.input['status']
+
+        activity.save()
+
+
+class ActivityMenu(APIView):
+
+    def get(self):
+        if not self.request.user.is_authenticated():
+            raise ValidateError('User has not logged in.')
+        activityListAll = Activity.objects.all()
+        wechat_menu = CustomWeChatView.lib.get_wechat_menu()
+        if len(wechat_menu) >= 2:
+            actiListInMenu = wechat_menu[1]['sub_button']
+        else:
+            actiListInMenu = []
+        data = []
+        for acti in activityListAll:
```

Comment:
Should use object filter to get all the activities you want from the database instead of filtering them in a loop.